Upgrades version of Stripe Checkout

ribordy opened the pull request:
This takes users off-domain to complete their payment, then sends them back to the donation page with a confirmation shown when they return after completing their payment.

https://www.loom.com/share/7a3cec28f01e42e393c08a017b1af141

~This is blocked on https://github.com/pizza-to-the-polls/pizzabase/pull/24, but once the API supports it, we should be good to switch over!~

jamesarosen commented on src/components/page-donate/page-donate.tsx:
```diff
@@ -19,54 +8,50 @@ export class PageDonate {
   public componentWillLoad() {
     document.title = `Donate | Pizza to the Polls`;
-  }
+    const urlParams = new URLSearchParams(window.location.search);
```
URLSearchParams has [no IE support](https://caniuse.com/?search=urlsearchparams). I don't know how much of the PttP audience that is, but probably not much. If it proves to be a problem, we can use [Polyfill.io](https://polyfill.io/) to polyfill the API.

adamdehaven replied:
@ribordy & @jamesarosen if you don't want to use a polyfill, you could use this instead:
``` js
/**
 * Return the value of a URL parameter
 * @param url String 
 * @param key String
 */
const getUrlParam = (url, key) => {
  var results = new RegExp("[\?&]" + key + "=([^&#]*)").exec(url);
  if (results == null){
     return null;
  }
  return decodeURI(results[1]) || 0;
};

// Get the URL
const url = window.location.href;

// Usage
if (getUrlParam(url, "success") && getUrlParam(url, "amount_usd")) {
  // Do something
}
```

ribordy replied:
Good catch! I didn't realize it didn't have total support. I had previously used the `query-string` npm package, but since this snippet is pretty straightforward, I'll just use it.

adamdehaven replied:
Wasn't sure what you were using for url so obviously leave out that `const url = ... ` or whatever you don't need

jamesarosen commented on src/components/page-donate/page-donate.tsx:
```diff
@@ -19,54 +8,50 @@ export class PageDonate {
-    let handler: any = null;
-    const StripeCheckout: any = (window as any).StripeCheckout;
+    if (urlParams.get("success") && urlParams.get("amount_usd")) {
+      this.amount = parseInt(urlParams.get("amount_usd") as string, 10);
```
I'm surprised the `as string` is necessary. The type of `urlParams.get(…)` should be `string | undefined` and TypeScript can tell that we're in an `if` block that excludes the `undefined` case.

Certainly not a problem though.

ribordy replied:
It doesn't seem like the linter (or compiler?) is smart enough for that--at least VSCode shows a red squiggly when I don't explicitly specify.

jamesarosen commented on src/components/page-donate/page-donate.tsx:
```diff
@@ -19,54 +8,50 @@ export class PageDonate {
+      headers: { "Content-Type": "application/json" },
+    });
+
+    if (resp.status === 200) {
```
Do you want any sort of error-handling?

ribordy replied:
Will add error handling here!

jamesarosen commented on src/components/page-donate/page-donate.tsx:
```diff
@@ -19,54 +8,50 @@ export class PageDonate {
+          .redirectToCheckout({
+            sessionId,
+          })
+          .then(function (result: any) {
+            console.error(result.error.message);
+          });
```
You have the `console.error` in the `then`. Did you mean `.catch`?

ribordy replied:
Lemme check on this--I wanna say I copied this from the docs; if `redirectToCheckout` succeeds, we won't be on this page anyway. Need to confirm how it does error handling.

ribordy replied:
The [docs](https://stripe.com/docs/js/checkout) seem to suggest that an error will be surface in this manner. Seems non-standard.

jamesarosen replied:
> if redirectToCheckout succeeds, we won't be on this page anyway

There's some nuance there. A `window.location = …` actually happens asynchronously, so the remainder of the current JS execution will finish before the page unloads.

```js
function foo() {
  console.log('this happens')

  window.location = 'https://example.com'

  console.log('this happens too')

  setTimeout(() => {
    console.log('this never happens')
  }, 0)
}
```

ribordy replied:
Oh huh, TIL!

pramodthammaiah commented on src/components/page-donate/page-donate.tsx:
```diff
@@ -19,54 +8,50 @@ export class PageDonate {
-    // Get referral from URL
-    const urlParams = new URLSearchParams(window.location.search);
-    const referral = urlParams.get("referral") || "";
+  public async donate(amount: number) {
+    const resp = await fetch(`${process.env.PIZZA_BASE_DOMAIN}/donations`, {
+      body: JSON.stringify({ amountUsd: amount, referrer: this.referral }),
+      method: "POST",
+      mode: "cors",
+      headers: { "Content-Type": "application/json" },
+    });
```
IIUC stripe will send a charge webhook to the donationsController with https://dashboard.stripe.com/webhooks @mojowen -- thoughts on having polls.pizza sending the ping vs getting it from Stripe? I don't think it's a big deal since the Donation entity has a unique check on the stripe charge id. Though it may double the hits to the donationController.

pramodthammaiah replied:
Nvm, didn't see the recent changes to DonationsController.

ribordy replied:
Yep--just for the record, this simply initiates the Stripe checkout. When the checkout succeeds, it looks like the plan is for Stripe to ping a separate endpoint on the Donations controller to register that payment with us.

jamesarosen commented on src/components/page-donate/page-donate.tsx:
```diff
@@ -19,54 +8,50 @@ export class PageDonate {
+      body: JSON.stringify({ amountUsd: amount, referrer: this.referral }),
+      method: "POST",
+      mode: "cors",
+      headers: { "Content-Type": "application/json" },
```
You might want to supply an `Accept` as well since you do `resp.json()` below.

ribordy commented:
I think I've addressed all the comments aside from user-facing error messaging. @mojowen do we have an example of how to surface these to the user elsewhere in the app?

I've added a `showError` method that we can fill in once we make a decision. Since the existing donate page doesn't have any error handling, I propose we give this a go on staging while we figure out next steps.

ribordy commented:
Added error text! 
![image](https://user-images.githubusercontent.com/2159914/97772815-c3491600-1b07-11eb-8ee1-30cf06536069.png)

mojowen commented:
Nice! Let me know when you're ready for a review

mojowen commented:
@ribordy I gave you write access if you want to open the branch up here and test on our staging subdomain https://next.polls.pizza - I only think that (and the other tests) when the branch on this repo

ribordy commented:
Yep, will close and re-open on this repo
